src/software/backend/replay_backend.cpp

```diff
@@ -0,0 +1,46 @@
+#include "replay_backend.h"
+
+#include "software/util/design_patterns/generic_factory.h"
+
+const std::string ReplayBackend::name = "replay";
+
+ReplayBackend::ReplayBackend(
+    std::shared_ptr<const ReplayBackendConfig> replay_backend_config)
+    : Backend(),
+      replay_reader(replay_backend_config->ReplayPath()->value()),
+      pull_from_replay_thread(
+          boost::bind(&ReplayBackend::continuouslyPullFromReplayFiles, this)),
+      last_msg_received_time(std::nullopt),
+      last_msg_replayed_time(std::nullopt)
+{
+}
+
+// do nothing
+// TODO: perhaps we want to encode the output primitives to proto and log them?
+void ReplayBackend::onValueReceived(ConstPrimitiveVectorPtr primitives) {}
+
+// do nothing
+void ReplayBackend::onValueReceived(World world) {}
+
+void ReplayBackend::continuouslyPullFromReplayFiles()
+{
+    while (auto sensor_msg_or_null = replay_reader.getNextMsg())
+    {
+        auto this_msg_received_time = std::chrono::duration<double>(
+            sensor_msg_or_null->time_received().epoch_timestamp_seconds());
+
+        if (last_msg_received_time && last_msg_replayed_time)
+        {
+            std::this_thread::sleep_until(
+                *last_msg_replayed_time +
+                (this_msg_received_time - *last_msg_received_time));
+        }
+        this->sendValueToObservers(*sensor_msg_or_null);
```

Comment:
```suggestion
        Subject<SensorMsg>::sendValueToObservers(*sensor_msg_or_null);
```

Reply:
I think SFINAE will result in the correct function being called?

Reply:
yeah I'm sure it's correct, but it's more clear to invoke the superclass. Not a blocking comment